app/src/test/java/de/westnordost/streetcomplete/data/user/oauth/OAuthAuthorizationTest.kt

```diff
@@ -0,0 +1,129 @@
+    }
+
+    @Test fun `generates different state for each instance`() {
+        val url1 = URL(createOAuth().createAuthorizationUrl())
+        val url2 = URL(createOAuth().createAuthorizationUrl())
+        assertTrue(url1.queryParameters["state"] != url2.queryParameters["state"])
+    }
+
+    @Test fun `serializes correctly`() {
+        val oauth1 = createOAuth()
+        val oauth1String = Json.encodeToString(oauth1)
+        val oauth2 = Json.decodeFromString<OAuthAuthorization>(oauth1String)
+        val oauth2String = Json.encodeToString(oauth2)
+
+        assertEquals(oauth1String, oauth2String)
+    }
+
+    @Test fun itsForMe() {
+        val oauth = createOAuth()
+        val state = urlEncode(URL(oauth.createAuthorizationUrl()).queryParameters["state"]!!)
+
+        assertFalse(oauth.itsForMe(URI("localhost://oauth"))) // no state
+        assertFalse(oauth.itsForMe(URI("localhost://oauth?state=abc"))) // different state
+        assertTrue(oauth.itsForMe(URI("localhost://oauth?state=$state"))) // same state
+        // different uri
+        assertFalse(oauth.itsForMe(URI("localhost://oauth3?state=$state")))
+        assertFalse(oauth.itsForMe(URI("localhost://oauth/path?state=$state")))
+        assertFalse(oauth.itsForMe(URI("localboost://oauth?state=$state")))
+    }
+
+    @Test fun `extractAuthorizationCode fails with useful error messages`() {
+        val oauth = createOAuth()
+
+        // server did not respond correctly with "error"
+        assertFailsWith<ConnectionException> {
+            oauth.extractAuthorizationCode(URI("localhost://oauth?e=something"))
+        }
+
+        try {
+            oauth.extractAuthorizationCode(URI("localhost://oauth?error=hey%2Bwhat%27s%2Bup"))
+        } catch (e: AuthorizationException) {
+            assertEquals("hey what's up", e.message)
+        }
+
+        try {
+            oauth.extractAuthorizationCode(URI("localhost://oauth?error=A%21&error_description=B%21"))
+        } catch (e: AuthorizationException) {
+            assertEquals("A!: B!", e.message)
+        }
+
+        try {
+            oauth.extractAuthorizationCode(URI("localhost://oauth?error=A%21&error_uri=http%3A%2F%2Fabc.de"))
+        } catch (e: AuthorizationException) {
+            assertEquals("A! (see http://abc.de)", e.message)
+        }
+
+        try {
+            oauth.extractAuthorizationCode(URI("localhost://oauth?error=A%21&error_description=B%21&error_uri=http%3A%2F%2Fabc.de"))
+        } catch (e: AuthorizationException) {
+            assertEquals("A!: B! (see http://abc.de)", e.message)
+        }
+    }
+
+    @Test fun extractAuthorizationCode() {
+        assertEquals(
+            "my code",
+            createOAuth().extractAuthorizationCode(URI("localhost://oauth?code=my%20code"))
+        )
+    }
+
+    // it's not properly possible to test retrieveAccessToken in isolation because the http client
+    // is not injected (passed in the constructor)
```

Comment:
Have you looked at https://wiremock.org/ ?
JOSM uses it to "mock" network responses for tests.

Reply:
I haven't, thanks for the link!